crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,361 @@
+        }
+    }
+}
+
+impl<CTX: ContextTr, P: PrecompileProvider<CTX, Output = InterpreterResult>> PrecompileProvider<CTX>
+    for MaybeCachedPrecompileProvider<P>
+{
+    type Output = P::Output;
+
+    fn set_spec(&mut self, spec: <CTX::Cfg as Cfg>::Spec) -> bool {
+        let old_spec = self.spec;
+        self.spec = spec.clone().into();
+
+        if self.precompile_provider.set_spec(spec) {
+            return true;
+        }
+
+        old_spec != self.spec
+    }
+
+    #[cfg(feature = "std")]
+    fn run(
+        &mut self,
+        context: &mut CTX,
+        address: &Address,
+        inputs: &InputsImpl,
+        is_static: bool,
+        gas_limit: u64,
+    ) -> Result<Option<Self::Output>, String> {
+        use revm::interpreter::{Gas, InstructionResult};
+
+        if let Some(cache) = &self.cache {
+            let key =
+                PrecompileKey { address: *address, spec: self.spec, input: inputs.input.clone() };
+
+            let cache_result = cache.cache.get(&key);
+
+            if let Some(ref entry) = cache_result {
+                // if gas_limit is below known lower bound, we know it will fail with OOG
+                if gas_limit <= entry.lower_gas_limit {
+                    self.increment_by_one_precompile_cache_hits();
+
+                    let result = InterpreterResult {
+                        result: InstructionResult::PrecompileOOG,
+                        gas: Gas::new(gas_limit),
+                        output: alloy_primitives::Bytes::new(),
+                    };
+
+                    return Ok(Some(result));
+                }
+
+                // if gas_limit is above upper bound, use the cached result
+                if gas_limit >= entry.upper_gas_limit {
+                    self.increment_by_one_precompile_cache_hits();
+
+                    // for successful results, we need to ensure gas costs are correct when
+                    // gas_limit differs
+                    if let Ok(mut result) = entry.result.clone() {
+                        if result.result == InstructionResult::Return {
+                            let mut adjusted_gas = Gas::new(gas_limit);
+                            adjusted_gas.set_spent(result.gas.spent());
+
+                            result.gas = adjusted_gas;
+                        }
+
+                        return Ok(Some(result));
+                    }
+
+                    return entry.result.clone().map(Some);
+                }
+            }
+
+            // call the precompile if cache miss
+            let output =
+                self.precompile_provider.run(context, address, inputs, is_static, gas_limit);
+
+            match &output {
+                Ok(Some(result)) => {
+                    self.increment_by_one_precompile_cache_misses();
+                    let previous_entry_count = self.cache_entry_count();
+
+                    if result.result == InstructionResult::PrecompileOOG {
```

Comment:
should we even cache errors?

Reply:
i think it can trade avoid execution in the best cases, maybe not worth it, lmk what do you think